plugins/experimental/carp/Common.cc

```diff
@@ -0,0 +1,364 @@
+        if (TS_SUCCESS == TSMimeHdrFieldAppend(reqp, hdr_loc, new_field)) {
+          TSDebug(DEBUG_TAG_HOOK, "\tAdded header %s: %s", header.c_str(), value.c_str());
+          bReturn = true;
+        }
+      }
+      TSHandleMLocRelease(reqp, hdr_loc, new_field);
+    }
+  }
+  return bReturn;
+}
+
+/************************************************************************/
+bool
+getHeader(TSMBuffer& reqp, TSMLoc& hdr_loc, const std::string header, std::string& value)
+{
+  bool bReturn = false;
+  TSMLoc fieldLoc = TSMimeHdrFieldFind(reqp, hdr_loc, header.data(), header.size());
+
+  if (fieldLoc && (fieldLoc != NULL)) {
+    const char *str;
+    int strLen = 0;
+    str = TSMimeHdrFieldValueStringGet(reqp, hdr_loc, fieldLoc, 0, &strLen);
+    if (str && strLen > 0) {
+      value.assign(str, strLen);
+      bReturn = true;
+    } else {
+      bReturn = false;
+    }
+    TSHandleMLocRelease(reqp, hdr_loc, fieldLoc);
+  }
+  return bReturn;
+}
+
+/************************************************************************/
+bool
+removeHeader(TSMBuffer& reqp, TSMLoc& hdr_loc, string header)
+{
+  bool bReturn = false;
+  TSMLoc fieldLoc = TSMimeHdrFieldFind(reqp, hdr_loc, header.data(), header.size());
+
+  if (fieldLoc && (fieldLoc != NULL)) {
+    if(TSMimeHdrFieldRemove(reqp, hdr_loc, fieldLoc) == TS_SUCCESS) {
+      if(TSMimeHdrFieldDestroy(reqp, hdr_loc, fieldLoc) == TS_SUCCESS) {
+        TSDebug(DEBUG_TAG_HOOK, "\tRemoved header %s", header.c_str());
+        bReturn = true;
+      }
+    }
+    TSHandleMLocRelease(reqp, hdr_loc, fieldLoc);
+  }
+  return bReturn;
+}
+
+
+bool
+setHeader(TSMBuffer& reqp, TSMLoc& hdr_loc, const std::string header, const std::string value)
+{
+  bool bReturn = false;
+  TSMLoc fieldLoc = TSMimeHdrFieldFind(reqp, hdr_loc, header.data(), header.size());
+
+  if (fieldLoc && (fieldLoc != NULL)) {
+    if (TSMimeHdrFieldValueStringSet(reqp, hdr_loc, fieldLoc, 0, value.data(), value.length()) == TS_SUCCESS) {
+      TSDebug(DEBUG_TAG_HOOK, "\tSet header %s to %s", header.c_str(), value.c_str());
+      bReturn = true;
+    } else {
+      bReturn = false;
+    }
+    TSHandleMLocRelease(reqp, hdr_loc, fieldLoc);
+  }
+  return bReturn;
+}
+
+/************************************************************************/
+//Convert a struct sockaddr address to a string, IPv4 and IPv6:
+bool
+getStringFromSockaddr(const struct sockaddr *sa, string& s)
+{
+  char str[INET6_ADDRSTRLEN];
+  memset(str,sizeof(str),0);
+  switch (sa->sa_family) {
+  case AF_INET:
+    inet_ntop(AF_INET, &(((struct sockaddr_in *) sa)->sin_addr), str, sizeof(str));
+    break;
+
+  case AF_INET6:
+    inet_ntop(AF_INET6, &(((struct sockaddr_in6 *) sa)->sin6_addr), str, sizeof(str));
+    break;
+
+  default:
+    s.assign("Unknown");
+    return false;
+  }
+  s.assign(str);
+  sprintf(str,":%d",ntohs(((struct sockaddr_in *) sa)->sin_port));
```

Comment:
You might want to use brackets around IPv6 addresses so the port colon and the internal IPv6 colons do not get mixed up.